reactivesocket-core/src/main/java/io/reactivesocket/ServerReactiveSocket.java

```diff
@@ -143,71 +147,75 @@ public ServerReactiveSocket start() {
-                            receiver.onComplete();
-                        }
-                        return Mono.empty();
-                    case ERROR:
-                        synchronized (ServerReactiveSocket.this) {
-                            receiver = receivers.get(streamId);
-                        }
-                        if (receiver != null) {
-                            receiver.onError(new ApplicationException(frame));
-                        }
-                        return Mono.empty();
-                    case NEXT_COMPLETE:
-                        synchronized (ServerReactiveSocket.this) {
-                            receiver = receivers.get(streamId);
-                        }
-                        if (receiver != null) {
-                            receiver.onNext(frame);
-                            receiver.onComplete();
-                        }
-
-                        return Mono.empty();
-
-                    case SETUP:
-                        return handleError(streamId, new IllegalStateException("Setup frame received post setup."));
-                    default:
-                        return handleError(streamId, new IllegalStateException("ServerReactiveSocket: Unexpected frame type: "
-                            + frame.getType()));
+                try {
+                    int streamId = frame.getStreamId();
+                    UnicastProcessor<Payload> receiver;
+                    switch (frame.getType()) {
+                        case FIRE_AND_FORGET:
+                            return handleFireAndForget(streamId, fireAndForget(new PayloadImpl(frame)));
+                        case REQUEST_RESPONSE:
+                            return handleRequestResponse(streamId, requestResponse(new PayloadImpl(frame)));
+                        case CANCEL:
+                            return handleCancelFrame(streamId);
+                        case KEEPALIVE:
+                            return handleKeepAliveFrame(frame);
+                        case REQUEST_N:
+                            return handleRequestN(streamId, frame);
+                        case REQUEST_STREAM:
+                            return handleStream(streamId, requestStream(new PayloadImpl(frame)), frame);
+                        case REQUEST_CHANNEL:
+                            return handleChannel(streamId, frame);
+                        case PAYLOAD:
+                            // TODO: Hook in receiving socket.
+                            return Mono.empty();
+                        case METADATA_PUSH:
+                            return metadataPush(new PayloadImpl(frame));
+                        case LEASE:
+                            // Lease must not be received here as this is the server end of the socket which sends leases.
+                            return Mono.empty();
+                        case NEXT:
+                            synchronized (ServerReactiveSocket.this) {
```

Comment:
I'd suggest pulling this out to a synchronized method to avoid the number of places that enforce this synchronization block